qiskit/transpiler/passes/synthesis/hls_plugins.py

```diff
@@ -1624,3 +1633,145 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
+    an :class:`~.HLSConfig` object to use this method with :class:`~.HighLevelSynthesis`.
+    """
+
+    def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **options):
+        # pylint: disable=cyclic-import
+        from .high_level_synthesis import _synthesize_operation
+
+        if not isinstance(high_level_object, AnnotatedOperation):
+            return None
+
+        operation = high_level_object
+        modifiers = high_level_object.modifiers
+
+        # The plugin needs additional information that is not yet passed via the run's method
+        # arguments: namely high-level-synthesis data and options, the global qubits over which
+        # the operation is defined, and the initial state of each global qubit.
+        tracker = options.get("qubit_tracker", None)
+        data = options.get("hls_data")
+        input_qubits = options.get("input_qubits")
+
+        if len(modifiers) > 0:
+            num_ctrl = sum(
+                mod.num_ctrl_qubits for mod in modifiers if isinstance(mod, ControlModifier)
+            )
+            total_power = sum(mod.power for mod in modifiers if isinstance(mod, PowerModifier))
+            is_inverted = sum(1 for mod in modifiers if isinstance(mod, InverseModifier)) % 2
+
+            # The base operation cannot use control qubits as auxiliary qubits.
+            # In addition, when we have power or inverse modifiers, we need to set all of
+            # the operation's qubits to dirty. Note that synthesizing the base operation we
+            # can use additional auxiliary qubits, however they would always be returned to
+            # their previous state, so clean qubits remain clean after each for- or while- loop.
+            annotated_tracker = tracker.copy()
+            annotated_tracker.disable(input_qubits[:num_ctrl])  # do not access control qubits
+            if total_power != 0 or is_inverted:
+                annotated_tracker.set_dirty(input_qubits)
+
+            # First, synthesize the base operation of this annotated operation.
+            # Note that synthesize_operation also returns the output qubits on which the
+            # operation is defined, however currently the plugin mechanism has no way
+            # to return these (and instead the upstream code greedily grabs some ancilla
+            # qubits from the circuit). We should refactor the plugin "run" iterface to
+            # return the actual ancilla qubits used.
+            synthesized_base_op, _ = _synthesize_operation(
+                operation.base_op, input_qubits[num_ctrl:], data, annotated_tracker
+            )
+
+            # The base operation does not need to be synthesized.
+            # For simplicity, we wrap the instruction into a circuit. Note that
+            # this should not deteriorate the quality of the result.
+            if synthesized_base_op is None:
+                synthesized_base_op = _instruction_to_circuit(operation.base_op)
```

Comment:
I'm wondering whether we can keep this a `Gate | Instruction`, all methods inside `apply_annotation` should also work then no? One less conversion 🙂